### PR TITLE
Move compiler from netcoreapp3.1 to net6.0

### PR DESCRIPTION
**Pull request by agocke**
By far the biggest change here is adding `using ReferenceEqualityComparer = Roslyn.Utilities.EqualityComparer;` to a lot of places. VS Code also corrected the format when I did that, so set ignore whitespace if that's getting in your way while reviewing.

### CODE REVIEW

**Comment by sharwell** [comment]
> VS Code also corrected the format when I did that

This appears to be a bug in your local VS Code configuration. The trailing whitespaces should not have been removed by the IDE in response to this change.

**Comment by agocke** [comment]
@sharwell Yeah, probably set that up a while ago. Not a bad thing to get rid of trailing whitespace, though.

**Comment by jaredpar** [comment]
I think you may need to rebase against head based on a PR I just hit the bug with.

**Comment by agocke** [comment]
This is ready for review

**Review by genlu — APPROVED** [review]
Approved on behalf of infraswat

File: src/Compilers/Core/Portable/DiaSymReader/SymUnmanagedFactory.cs [hunk 1]
```diff
@@ -154,6 +154,9 @@ private static Type GetComTypeType(ref Type lazyType, Guid clsid)
         {
             if (lazyType == null)
             {
+#if NET6_0_OR_GREATER
+                System.Diagnostics.Debug.Assert(OperatingSystem.IsWindows());
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiaSymReader/SymUnmanagedFactory.cs:158]: @agocke why did you add this pattern in so many places?

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 2]
```diff
@@ -29,6 +29,7 @@
 using Microsoft.CodeAnalysis.Symbols;
 using Microsoft.DiaSymReader;
 using Roslyn.Utilities;
+using ReferenceEqualityComparer = Roslyn.Utilities.ReferenceEqualityComparer;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:32]: Longer term should we possibly rename this to `RoslynReferenceEqualityComparer` to avoid this using everywhere? Or maybe just unify on the BCL version? @sharwell, @RikkiGibson

Comment (RikkiGibson) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:32]: BCL version looks like it would behave the same as ours https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ReferenceEqualityComparer.cs

Comment (RikkiGibson) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:32]: IReadOnlySet we would want to keep, though, I guess. Ours has way fewer members on it than the platform version.

File: src/Compilers/Core/Portable/DiaSymReader/SymUnmanagedFactory.cs (hunk 1 above)
Comment (agocke) [reply src/Compilers/Core/Portable/DiaSymReader/SymUnmanagedFactory.cs:158]: Sorry, missed this comment. This is a new analyzer in 6.0. The calls are annotated as "Windows only", so I asserted that we're in Windows to make the warnings go away. Flowing them upwards would also work, but it gets tricky.
